File: protocol/src/main/java/spp/protocol/platform/error/EventBusUtil.java
package spp.protocol.platform.error;

import spp.protocol.probe.error.LiveInstrumentException;

import static spp.protocol.probe.error.LiveInstrumentException.ErrorType;

public class EventBusUtil {

    public static Exception fromEventBusException(String exception) {
        if (exception.startsWith("EventBusException")) {
            String exceptionType = substringAfter(exception, "EventBusException:");
            exceptionType = substringBefore(exceptionType, "[");
            String exceptionParams = substringAfter(exception, "[");
            exceptionParams = substringBefore(exceptionParams, "]");
            String exceptionMessage = substringAfter(exception, "]: ").trim();

            if (LiveInstrumentException.class.getSimpleName().equals(exceptionType)) {
                return new LiveInstrumentException(ErrorType.valueOf(exceptionParams), exceptionMessage)
                        .toEventBusException();
            } else {
                throw new UnsupportedOperationException(exceptionType);
            }
        } else {
            throw new IllegalArgumentException(exception);
        }
    }

    private static String substringAfter(String value, String delimiter) {
        int index = value.indexOf(delimiter);
        if (index == -1) return value;
        else return value.substring(index + delimiter.length());
    }

    private static String substringBefore(String value, String delimiter) {
        int index = value.indexOf(delimiter);
        if (index == -1) return value;
        else return value.substring(0, index);
    }
}

File: protocol/src/main/kotlin/spp/protocol/platform/error/EventBusUtil.kt
package spp.protocol.platform.error

import spp.protocol.probe.error.LiveInstrumentException

object EventBusUtil {

    fun fromEventBusException(exception: String): Exception {
        return if (exception.startsWith("EventBusException")) {
            var exceptionType = exception.substringAfter("EventBusException:")
            exceptionType = exceptionType.substringBefore("[")
            var exceptionParams = exception.substringAfter("[")
            exceptionParams = exceptionParams.substringBefore("]")
            val exceptionMessage = exception.substringAfter("]: ").trim { it <= ' ' }
            if (LiveInstrumentException::class.java.simpleName == exceptionType) {
                LiveInstrumentException(
                    LiveInstrumentException.ErrorType.valueOf(exceptionParams),
                    exceptionMessage
                ).toEventBusException()
            } else {
                throw UnsupportedOperationException(exceptionType)
            }
        } else {
            throw IllegalArgumentException(exception)
        }
    }
}
